// @rsdoctor/utils@1.1.2/dist/esm/common/lodash.js
function isUndefined(value) {
  return typeof value === "undefined";
}
function isNumber(value) {
  return typeof value === "number" && !Number.isNaN(value);
}
function isObject(value) {
  return typeof value === "object" && value !== null;
}
function isEmpty(value) {
  return value == null || // Check for null or undefined
  Array.isArray(value) && value.length === 0 || // Check for empty array
  typeof value === "object" && Object.keys(value).length === 0;
}
function last(array) {
  return array[array.length - 1];
}
function compact(array) {
  return array.filter((item) => item != null || !item);
}
function isNil(value) {
  return value === null || value === void 0;
}
const isPlainObject = (obj) => {
  return obj !== null && typeof obj === "object" && Object.getPrototypeOf(obj) === Object.prototype;
};
const isString = (v) => typeof v === "string" || !!v && typeof v === "object" && !Array.isArray(v) && {}.toString.call(v) === "[object String]";
function pick(obj, keys) {
  const result = {};
  for (let i = 0; i < keys.length; i++) {
    const key = keys[i];
    if (Object.hasOwn(obj, key)) {
      result[key] = obj[key];
    }
  }
  return result;
}
export {
  compact,
  isEmpty,
  isNil,
  isNumber,
  isObject,
  isPlainObject,
  isString,
  isUndefined,
  last,
  pick
};
