// @lynx-js/web-constants@0.10.1/dist/endpoints.js
// Copyright 2023 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.
import { createRpcEndpoint } from '@lynx-js/web-worker-rpc';
export const postExposureEndpoint = createRpcEndpoint('__postExposure', false, false);
export const publicComponentEventEndpoint = createRpcEndpoint('publicComponentEvent', false, false);
export const publishEventEndpoint = createRpcEndpoint('publishEvent', false, false);
export const postOffscreenEventEndpoint = createRpcEndpoint('postOffscreenEventEndpoint', false, false);
export const switchExposureServiceEndpoint = createRpcEndpoint('switchExposureServiceEndpoint', false, false);
export const mainThreadStartEndpoint = createRpcEndpoint('mainThreadStart', false, false);
export const updateDataEndpoint = createRpcEndpoint('updateData', false, true);
export const sendGlobalEventEndpoint = createRpcEndpoint('sendGlobalEventEndpoint', false, false);
export const disposeEndpoint = createRpcEndpoint('dispose', false, true);
export const uiThreadFpReadyEndpoint = createRpcEndpoint('uiThreadFpReady', false, false);
export const onLifecycleEventEndpoint = createRpcEndpoint('__OnLifecycleEvent', false, false);
export const BackgroundThreadStartEndpoint = createRpcEndpoint('start', false, true);
/**
 * Error message, info
 */
export const reportErrorEndpoint = createRpcEndpoint('reportError', false, false);
export const flushElementTreeEndpoint = createRpcEndpoint('flushElementTree', false, true);
export const mainThreadChunkReadyEndpoint = createRpcEndpoint('mainThreadChunkReady', false, false);
export const callLepusMethodEndpoint = createRpcEndpoint('callLepusMethod', false, true);
export const invokeUIMethodEndpoint = createRpcEndpoint('__invokeUIMethod', false, true);
export const setNativePropsEndpoint = createRpcEndpoint('__setNativeProps', false, true);
export const nativeModulesCallEndpoint = createRpcEndpoint('nativeModulesCall', false, true);
export const napiModulesCallEndpoint = createRpcEndpoint('napiModulesCall', false, true, true);
export const getCustomSectionsEndpoint = createRpcEndpoint('getCustomSections', false, true);
export const markTimingEndpoint = createRpcEndpoint('markTiming', false, false);
export const postTimingFlagsEndpoint = createRpcEndpoint('postTimingFlags', false, false);
export const triggerComponentEventEndpoint = createRpcEndpoint('__triggerComponentEvent', false, false);
export const selectComponentEndpoint = createRpcEndpoint('__selectComponent', false, true);
export const dispatchLynxViewEventEndpoint = createRpcEndpoint('dispatchLynxViewEvent', false, true);
export const dispatchNapiModuleEndpoint = createRpcEndpoint('dispatchNapiModule', false, false);
