// @lynx-js/web-elements@0.5.4/dist/XRefreshView/XRefreshViewEventsEmitter.js
import { __esDecorate, __runInitializers, __setFunctionName } from "tslib";
/*
// Copyright 2024 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.
*/
import { genDomGetter, } from '@lynx-js/web-elements-reactive';
import { XRefreshIntersectionObserverEvent } from './XRefreshSubElementIntersectionObserver.js';
import { commonComponentEventSetting } from '../common/commonEventInitConfiguration.js';
import { registerEventEnableStatusChangeHandler } from '@lynx-js/web-elements-reactive';
let XRefreshViewEventsEmitter = (() => {
    let _instanceExtraInitializers = [];
    let _private_handleComplexEventEnableAttributes_decorators;
    let _private_handleComplexEventEnableAttributes_descriptor;
    let _private_handleXEnableHeaderOffsetEvent_decorators;
    let _private_handleXEnableHeaderOffsetEvent_descriptor;
    return class XRefreshViewEventsEmitter {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
            _private_handleComplexEventEnableAttributes_decorators = [registerEventEnableStatusChangeHandler('headeroffset'), registerEventEnableStatusChangeHandler('headershow'), registerEventEnableStatusChangeHandler('footeroffset')];
            _private_handleXEnableHeaderOffsetEvent_decorators = [registerEventEnableStatusChangeHandler('startrefresh'), registerEventEnableStatusChangeHandler('headerreleased'), registerEventEnableStatusChangeHandler('startloadmore'), registerEventEnableStatusChangeHandler('footerreleased')];
            __esDecorate(this, _private_handleComplexEventEnableAttributes_descriptor = { value: __setFunctionName(function (status, eventName) {
                    this
                        .#eventSwitches[eventName] = status;
                    const { headeroffset, headershow, footeroffset } = this.#eventSwitches;
                    if (headeroffset
                        || headershow
                        || footeroffset) {
                        this.#enableComplexRefreshViewEvents();
                    }
                    else {
                        this.#disableComplexRefreshViewEvents();
                    }
                }, "#handleComplexEventEnableAttributes") }, _private_handleComplexEventEnableAttributes_decorators, { kind: "method", name: "#handleComplexEventEnableAttributes", static: false, private: true, access: { has: obj => #handleComplexEventEnableAttributes in obj, get: obj => obj.#handleComplexEventEnableAttributes }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(this, _private_handleXEnableHeaderOffsetEvent_descriptor = { value: __setFunctionName(function (status, eventName) {
                    this
                        .#eventSwitches[eventName] = status;
                    const { startrefresh, headerreleased, startloadmore, footerreleased } = this.#eventSwitches;
                    if (headerreleased
                        || footerreleased
                        || startloadmore
                        || startrefresh) {
                        this.#enableSimpleRefreshViewEvents();
                    }
                    else {
                        this.#disableSimpleRefreshViewEvents();
                    }
                }, "#handleXEnableHeaderOffsetEvent") }, _private_handleXEnableHeaderOffsetEvent_decorators, { kind: "method", name: "#handleXEnableHeaderOffsetEvent", static: false, private: true, access: { has: obj => #handleXEnableHeaderOffsetEvent in obj, get: obj => obj.#handleXEnableHeaderOffsetEvent }, metadata: _metadata }, null, _instanceExtraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        #dom = __runInitializers(this, _instanceExtraInitializers);
        static observedAttributes = [];
        #getXRefreshHeader = genDomGetter(() => this.#dom, 'x-refresh-view > x-refresh-header:first-of-type');
        #getXRefreshFooter = genDomGetter(() => this.#dom, 'x-refresh-view > x-refresh-footer:first-of-type');
        constructor(dom) {
            this.#dom = dom;
            this.#dom.addEventListener(XRefreshIntersectionObserverEvent.EventName, this.#handleSubElementObserverEvent);
        }
        #eventSwitches = {
            headeroffset: false,
            headerreleased: false,
            startrefresh: false,
            footeroffset: false,
            headershow: false,
            footerreleased: false,
            startloadmore: false,
        };
        // complex events switches
        get #handleComplexEventEnableAttributes() { return _private_handleComplexEventEnableAttributes_descriptor.value; }
        get #handleXEnableHeaderOffsetEvent() { return _private_handleXEnableHeaderOffsetEvent_descriptor.value; }
        /**
         * handle header/footer showing events
         */
        #headerShowing = false;
        #headerFullyShown = false;
        #footerShowing = false;
        #footerFullyShown = false;
        #handleSubElementObserverEvent = (e) => {
            e.stopPropagation();
            if (e.target.tagName === 'X-REFRESH-HEADER') {
                this.#headerShowing = e.startShowing;
                this.#headerFullyShown = e.fullyShowing;
            }
            else {
                this.#footerShowing = e.startShowing;
                this.#footerFullyShown = e.fullyShowing;
            }
        };
        /**
         * Event without dragging info;
         */
        #simpleRefreshViewEventsEnabled = false;
        #enableSimpleRefreshViewEvents() {
            if (this.#simpleRefreshViewEventsEnabled)
                return;
            this.#dom.addEventListener('touchend', this.#handleTouchEndForEvent);
            this.#simpleRefreshViewEventsEnabled = true;
        }
        #handleTouchEndForEvent = () => {
            if (this.#headerFullyShown) {
                this.#dom.dispatchEvent(new CustomEvent('headerreleased', commonComponentEventSetting));
                this.#dom.dispatchEvent(new CustomEvent('startrefresh', {
                    ...commonComponentEventSetting,
                    detail: { isManual: this.#dom._nextRefreshIsManual },
                }));
                this.#dom._nextRefreshIsManual = true;
            }
            else if ((this.#dom.getAttribute('enable-auto-loadmore') === 'true'
                && this.#footerShowing)
                || this.#footerFullyShown) {
                this.#dom.dispatchEvent(new CustomEvent('footerreleased', commonComponentEventSetting));
                this.#dom.dispatchEvent(new CustomEvent('startloadmore', commonComponentEventSetting));
            }
        };
        #disableSimpleRefreshViewEvents() {
            if (this.#simpleRefreshViewEventsEnabled) {
                this.#dom.removeEventListener('touchend', this.#handleTouchEndForEvent);
            }
        }
        /**
         * Event with dragging info
         */
        #dragging = false;
        #complexRefreshViewEventEnabled = false;
        #enableComplexRefreshViewEvents() {
            if (this.#complexRefreshViewEventEnabled)
                return;
            this.#dom.addEventListener('touchstart', this.#handleTouchStartForDraggingStatus);
            this.#dom.addEventListener('touchend', this.#handleTouchEndForDraggingStatus);
            this.#dom.addEventListener('touchcancel', this.#handleTouchEndForDraggingStatus);
            this.#dom
                .shadowRoot.querySelector('#container')
                .addEventListener('scroll', this.#handleScroll);
        }
        #handleTouchEndForDraggingStatus = () => {
            this.#dragging = false;
        };
        #handleTouchStartForDraggingStatus = () => {
            this.#dragging = true;
        };
        #handleScroll = () => {
            if (this.#headerShowing
                && (this.#eventSwitches.headershow || this.#eventSwitches.headeroffset)) {
                const header = this.#getXRefreshHeader();
                if (header) {
                    const height = parseFloat(getComputedStyle(header).height);
                    const scrollTop = this.#dom.shadowRoot.querySelector('#container').scrollTop;
                    this.#dom.dispatchEvent(new CustomEvent('headershow', {
                        ...commonComponentEventSetting,
                        detail: {
                            isDragging: this.#dragging,
                            offsetPercent: 1 - scrollTop / height,
                        },
                    }));
                    this.#dom.dispatchEvent(new CustomEvent('headeroffset', {
                        ...commonComponentEventSetting,
                        detail: {
                            isDragging: this.#dragging,
                            offsetPercent: 1 - scrollTop / height,
                        },
                    }));
                }
            }
            else if (this.#footerShowing && this.#eventSwitches.footeroffset) {
                const footer = this.#getXRefreshFooter();
                if (footer) {
                    const contentDom = this.#dom.shadowRoot.querySelector('#container');
                    const scrollTop = contentDom.scrollTop;
                    const scrollHeight = contentDom.scrollHeight;
                    const height = parseFloat(getComputedStyle(footer).height);
                    this.#dom.dispatchEvent(new CustomEvent('footeroffset', {
                        ...commonComponentEventSetting,
                        detail: {
                            isDragging: this.#dragging,
                            offsetPercent: 1 - scrollTop / height,
                        },
                    }));
                }
            }
        };
        #disableComplexRefreshViewEvents() {
            if (this.#complexRefreshViewEventEnabled) {
                this.#dom.removeEventListener('touchstart', this.#handleTouchStartForDraggingStatus);
                this.#dom.removeEventListener('touchend', this.#handleTouchEndForDraggingStatus);
                this.#dom.removeEventListener('touchcancel', this.#handleTouchEndForDraggingStatus);
                this.#dom
                    .shadowRoot.querySelector('#container')
                    .removeEventListener('scroll', this.#handleScroll);
            }
        }
        dispose() {
            this.#disableSimpleRefreshViewEvents();
            this.#disableComplexRefreshViewEvents();
        }
    };
})();
export { XRefreshViewEventsEmitter };
