// @rsdoctor/utils@1.1.2/dist/esm/logger.js
import c from "picocolors";
import { logger } from "rslog";
import { Constants } from "@rsdoctor/types";
function debug(getMsg, prefix = "") {
  if (!process.env[Constants.RsdoctorProcessEnvDebugKey]) {
    return;
  }
  logger.level = "verbose";
  logger.debug(`${prefix} ${getMsg()}`);
}
logger.override({
  log: (message) => {
    console.log(`[Rsdoctor log] ${message}`);
  },
  info: (message) => {
    console.log(`[Rsdoctor info] ${message}`);
  },
  warn: (message) => {
    console.warn(`[Rsdoctor warn] ${message}`);
  },
  start: (message) => {
    console.log(`[Rsdoctor start] ${message}`);
  },
  ready: (message) => {
    console.log(`[Rsdoctor ready] ${message}`);
  },
  error: (message) => {
    console.error(`[Rsdoctor error] ${message}`);
  },
  success: (message) => {
    console.error(`[Rsdoctor success] ${message}`);
  },
  debug: (message) => {
    if (process.env.DEBUG) {
      console.log(`[Rsdoctor debug] ${message}`);
    }
  }
});
export {
  c as chalk,
  debug,
  logger
};
