// @rsdoctor/graph@1.1.2/dist/cjs/graph/chunk-graph/asset.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var asset_exports = {};
__export(asset_exports, {
  Asset: () => Asset
});
module.exports = __toCommonJS(asset_exports);
var import_types = require("@rsdoctor/types");
var import_node_zlib = require("node:zlib");
let id = 1;
class Asset {
  static init() {
    id = 1;
  }
  constructor(path, size, chunks, content) {
    this.id = id++;
    this.path = path;
    this.size = size;
    this.chunks = chunks;
    this.content = content;
  }
  toData(types) {
    return {
      id: this.id,
      path: this.path,
      size: this.size,
      gzipSize: this.gzipSize,
      chunks: this.chunks?.map((ck) => ck.id),
      content: types === import_types.SDK.ToDataType.NoSourceAndAssets || types === import_types.SDK.ToDataType.NoCode ? "" : this.content
    };
  }
  /* native asset */
  setChunks(chunks) {
    this.chunks = chunks;
  }
  setId(id2) {
    this.id = id2;
  }
  setGzipSize(content) {
    this.gzipSize = (0, import_node_zlib.gzipSync)(content, { level: 9 }).length;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Asset
});
