// @lynx-js/web-worker-runtime@0.10.1/dist/backgroundThread/background-apis/createPerformanceApis.js
// Copyright 2023 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.
export function createPerformanceApis(timingSystem) {
    let inc = 0;
    const performanceApis = {
        generatePipelineOptions: () => {
            const newPipelineId = `_pipeline_` + (inc++);
            return {
                pipelineID: newPipelineId,
                needTimestamps: false,
            };
        },
        onPipelineStart: function () {
            // Do nothing
        },
        markPipelineTiming: function (pipelineId, timingKey) {
            timingSystem.markTimingInternal(timingKey, pipelineId);
        },
        bindPipelineIdWithTimingFlag: function (pipelineId, timingFlag) {
            if (!timingSystem.pipelineIdToTimingFlags.has(pipelineId)) {
                timingSystem.pipelineIdToTimingFlags.set(pipelineId, []);
            }
            const timingFlags = timingSystem.pipelineIdToTimingFlags.get(pipelineId);
            timingFlags.push(timingFlag);
        },
    };
    return performanceApis;
}
