// @lynx-js/web-worker-runtime@0.10.1/dist/backgroundThread/background-apis/createNativeModules.js
// Copyright 2023 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.
import { nativeModulesCallEndpoint, switchExposureServiceEndpoint, } from '@lynx-js/web-constants';
export async function createNativeModules(uiThreadRpc, mainThreadRpc, nativeModulesMap) {
    const switchExposure = mainThreadRpc.createCall(switchExposureServiceEndpoint);
    const nativeModulesCall = uiThreadRpc.createCall(nativeModulesCallEndpoint);
    const lynxExposureModule = {
        resumeExposure() {
            switchExposure(true, true);
        },
        stopExposure(param) {
            switchExposure(false, param.sendEvent ?? true);
        },
    };
    const bridgeModule = {
        call(name, data, callback) {
            nativeModulesCall(name, data, 'bridge').then(callback);
        },
    };
    const nativeModules = {};
    const customNativeModules = {};
    await Promise.all(Object.entries(nativeModulesMap).map(([moduleName, moduleStr]) => import(/* webpackIgnore: true */ moduleStr).then(module => customNativeModules[moduleName] = module?.default?.(nativeModules, (name, data) => nativeModulesCall(name, data, moduleName)))));
    return Object.assign(nativeModules, {
        bridge: bridgeModule,
        LynxExposureModule: lynxExposureModule,
        ...customNativeModules,
    });
}
