// @lynx-js/offscreen-document@0.0.0/dist/webworker/OffscreenCSSStyleDeclaration.js
// Copyright 2023 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.
import { ancestorDocument } from './OffscreenElement.js';
import { operations } from './OffscreenDocument.js';
import { OperationType } from '../types/ElementOperation.js';
import { uniqueId } from './OffscreenNode.js';
export class OffscreenCSSStyleDeclaration {
    /**
     * @private
     */
    _parent;
    constructor(parent) {
        this._parent = parent;
    }
    setProperty(property, value, priority) {
        this._parent[ancestorDocument][operations].push({
            type: OperationType['StyleDeclarationSetProperty'],
            uid: this._parent[uniqueId],
            property,
            value: value,
            priority: priority,
        });
    }
    removeProperty(property) {
        this._parent[ancestorDocument][operations].push({
            type: OperationType['StyleDeclarationRemoveProperty'],
            uid: this._parent[uniqueId],
            property,
        });
    }
}
